File: LeetCodePractice/src/DP/BestTimeToBuyAndSellStockWithCooldown.java
package DP;

/*Say you have an array for which the ith element is the price of a given stock on day i.

Design an algorithm to find the maximum profit. You may complete as many transactions as you like (ie, buy one and sell one share of the stock multiple times) with the following restrictions:

You may not engage in multiple transactions at the same time (ie, you must sell the stock before you buy again).
After you sell your stock, you cannot buy stock on next day. (ie, cooldown 1 day)
Example:

Input: [1,2,3,0,2]
Output: 3 
Explanation: transactions = [buy, sell, cooldown, buy, sell]*/
public class BestTimeToBuyAndSellStockWithCooldown {
	public int maxProfit(int[] prices) {
		if (prices == null || prices.length == 0) {
			return 0;
		}
		int n = prices.length;
		int[] buy = new int[n + 1];
		int[] sell = new int[n + 1];
		int[] rest = new int[n + 1];
		buy[0] = Integer.MIN_VALUE;
		for (int i = 1; i <= n; i++) {
			// buy[i] means max profit when day i is in buy state, 
			//ex: buy, buy, buy, sell, rest, buy, buy
			buy[i] = Math.max(buy[i - 1], rest[i - 1] - prices[i - 1]);
			sell[i] = Math.max(sell[i - 1], buy[i - 1] + prices[i - 1]);
			rest[i] = Math.max(rest[i - 1], buy[i - 1]);
			rest[i] = Math.max(rest[i], sell[i - 1]);
		}
		int res = Math.max(sell[n], rest[n]);
		return res;

	}
}

File: LeetCodePython/BestTimeToBuyAndSellStockWithCooldown.py
"""
You are given an array prices where prices[i] is the price of a given stock on the ith day.

Find the maximum profit you can achieve. You may complete as many transactions as you like (i.e., buy one and sell one share of the stock multiple times) with the following restrictions:

After you sell your stock, you cannot buy stock on the next day (i.e., cooldown one day).
Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).

Example 1:
Input: prices = [1,2,3,0,2]
Output: 3
Explanation: transactions = [buy, sell, cooldown, buy, sell]

Example 2:
Input: prices = [1]
Output: 0

Constraints:
1 <= prices.length <= 5000
0 <= prices[i] <= 1000
"""
from typing import List


class BestTimeToBuyAndSellStockWithCooldown:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        bought, sold, rest = [0] * (n + 1), [0] * (n + 1), [0] * (n + 1)
        bought[0] = float('-inf')
        for i in range(n):
            bought[i + 1] = max(bought[i], rest[i] - prices[i])
            sold[i + 1] = max(sold[i], bought[i] + prices[i])
            rest[i + 1] = max(rest[i], bought[i], sold[i])
        return max(sold[-1], rest[-1])